### ISSUE TITLE
PROTEUS grid-search or forward model

### ISSUE THREAD

**Issue by timlichtenberg** [FormingWorlds/PROTEUS#204]
For moving towards an inverse method of PROTEUS sometime down the road, we need to consider a computationally feasible approach to run many models to fit a given set of observations.

To give an example of the problem: Let's assume a given exoplanet has the following known/observed parameters with uncertainties: stellar age, orbital distance, planet radius, planet mass, transmission/emission spectrum. Given these parameters, we would like to compute the best-fitting PROTEUS models over a set of input parameters, and then compute a goodness-of-fit metric. This is essentially the description of an atmospheric retrieval, only that PROTEUS simulations are way too computationally expensive as to perform 100k+ simulations.

I am not certain yet what is the best strategy to approach this problem. Here are a few that have some opportunities and drawbacks:
- A modified chi-squared or [r-squared](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html) algorithm to compute some measure of goodness-of-fit for an arbitrary grid. E.g. [Madhusudhan & Seager (2009)](https://iopscience.iop.org/article/10.1088/0004-637X/707/1/24/pdf).
- Train a machine learning model on the simulation data and use the machine learning model for the retrieval, e.g., [Ardevol Martinez et al. (2024)](https://www.aanda.org/articles/aa/abs/2024/01/aa48367-23/aa48367-23.html).
- Brute-force retrieval approach, e.g., nested sampling, MCMC, or some other random sampler, ignoring the computational cost, and possibly achieving a pretty low-confidence result.

**Comment by timlichtenberg** [comment]
The machine learning method of Ardevol Martinez (2024) seems potentially useful to me for this in principle. It can work together with an arbitrary "atmospheric code" (which could be PROTEUS) and does not need a large amount of training data to be generated upfront. Code here: https://github.com/franciscoardevolmartinez/FlopPITy.

**Comment by nichollsh** [comment]
I agree that this would be incredibly powerful. I can imagine that running an MCMC (or similar method) would be tricky because of the slow runtimes. When we are ready to look into this, maybe we could involve someone who has experience doing retrievals with large models?

**Comment by timlichtenberg** [comment]
I believe they need fewer simulations than a "normal" Bayesian model, which is one of their selling points. Nevertheless, even 100k simulations are not impossible when using a large-scale computing facility. We can and should do this sometime in the next year to achieve a large simulation grid, once the current plans with aragog and zephyrus are done. Cosmology solves this problem by running updated large-scale forward models every few years with high-performance codes (e.g. [Illustris project](https://www.illustris-project.org/)) and then using these models to train machine learning on them. This is a way to go, but if we can find an algorithm that enables running highly specialised simulations in a nexted sampling way and compute the Bayesian evidence directly, this would be preferable I think.

**Comment by timlichtenberg** [comment]
This is pretty cool! I like it a lot, this seems like a good solution to analyse our outputs. However, it still requires us to set the parameter space by hand, which means that we run many models that are not particularly useful/necessary and do not add valuable information. A Bayesian model selection or sth similar would save computation time until a statistically robust answer is achieved.

**Comment by stefsmeets** [comment]
This tool came up today: https://wandb.ai/

It's meant for hyperparameter optimization in machine learning. Via a [yaml file](https://docs.wandb.ai/guides/sweeps/define-sweep-configuration) you define the parameters to search. You just have to write the [interface in python](https://docs.wandb.ai/guides/track/launch).

**Comment by nichollsh** [comment]
Also on the theme of optimisation, I think we should consider Emcee, since it's well established within the astronomy community.

https://emcee.readthedocs.io/en/stable/

**Comment by nichollsh** [comment]
@timlichtenberg should this issue be closed in favour of Emma's PR #423 and Ben's BO work #162 ?
